### PR TITLE
Spawn a task in `oximeter` to actually do the collection

### PR DESCRIPTION
**Pull request by bnaecker**
`oximeter` currently starts up a task for each producer. When their collection interval expires, that task directly makes an HTTP request to the producer to collect its data, _in-line_ in the same task. This can cause problems, especially if there are concurrent updates to the producer's address information. The main `oximeter` HTTP server will send a message to that task in that case, asking it to update its producer information, but that task might be off in lalaland making an HTTP request. This is particularly bad since that HTTP request may never complete -- the producer's address might be updated, so the old one is defunct, possibly forever!

This commit addresses this by spawning a new task to actually run the collection itself. This keeps the first-level task alive and responsive, such that it can abort any previously-spawned collections if there are updates. We _also_ replace any existing task if a collection was explicitly requested, but we ignore existing collections (rather than spawning a new one) if the timer expires while one is already running.

This resolves test flakes described in #6901.

### CODE REVIEW

**Comment by bnaecker** [comment]
Some of the self-stat counter tests are failing, e.g., [here](https://buildomat.eng.oxide.computer/wg/0/details/01JD0TSYJD7E77H6GFBYKVECCD/vcoQvUeiE3kbCW7U6WA9KJ4D1fl1ghKW2FSSHvRHLIoJu0tF/01JD0TTEFK6P7899R18DJ8VBJ8#S6284). These have always been a bit flaky, so I'm not that surprised that adding another layer of spawning has made them more so. I've set up a small run-loop locally, which just runs these failing tests in a loop until they fail. I can get them both to fall over within about a minute, usually around 60 runs.

File: oximeter/collector/src/agent.rs [hunk 1]
```diff
@@ -131,29 +125,153 @@ async fn perform_collection(
...
+        match self.fut.as_mut() {
+            Some(handle) => {
+                match handle.as_mut().poll(cx) {
+                    Poll::Ready(res) => {
+                        // NOTE: Take out of the handle too.
+                        //
+                        // This is important. `JoinHandle` will panic if we poll
+                        // it after completion, so remove it here so that we
+                        // never enter this match arm at all if we have no
+                        // handle, on the next time we're polled.
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply oximeter/collector/src/agent.rs:179]: I agree that this is awkward, but it's not really a "single future being ready multiple times". This is similar to [`OptionFuture`](https://docs.rs/futures/latest/futures/future/struct.OptionFuture.html), but probably less well-thought out. I'm happy to go a different direction, using some channels to control a spawned collection rather than this method.

**Comment by bnaecker** [comment]
@jgallagher and I chatted at length about this. I'm going to push some big changes to this branch, but the summary is:

- spawn another long-running task to do the collections
- that task waits for either a timer-triggered or explicit collection request, both are treated the same. These are queued by the first task, with a pretty small queue-size limit (1, even).
- once that task receives a request for work, it starts a `select!` loop that: waits for the work / collection to complete or timeout, _and_ waits for updates to the producer endpoint information. The latter is forwarded in a `watch` channel from the first task, since we only care about the latest value here and don't care if anyone is explicitly listening.
- If this task gets an update while a collection is running, it cancels and recreates the collection with the updated information. This is to avoid returning an annoying error that says, "I have all the information I need to retry your request, but I'm going to make _you_ do it instead."

File: oximeter/collector/src/agent.rs [hunk 2]
```diff
@@ -166,35 +295,112 @@ async fn collection_task(
...
+                            Err(e) => {
+                                match e {
+                                    TrySendError::Closed(tok) => {
+                                        debug!(
+                                            log,
+                                            "collection task forced collection \
+                                            queue is closed. Attempting to \
+                                            notify caller and exiting.",
+                                        );
+                                        let _ = tok.send(Ok(()));
```
Comment (jgallagher) [inline oximeter/collector/src/agent.rs:355]: Should this be some kind of error instead of `Ok(())`? In this case we definitely did not force a collection to happen, right?

File: oximeter/collector/src/agent.rs [hunk 3]
```diff
@@ -166,35 +295,112 @@ async fn collection_task(
...
                         );
 
-                        // Update the logger with the new information as well.
-                        log = orig_log.new(o!("address" => producer.address));
-                        collection_timer = interval(producer.interval);
+                        // Update our own collection timer, and notify the
+                        // collection task of the new data as well.
+                        collection_timer = interval(new_info.interval);
                         collection_timer.tick().await; // completes immediately
+                        if let Err(_) = producer_info_tx.send(new_info) {
```
Comment (jgallagher) [inline oximeter/collector/src/agent.rs:396]: To combine this a bit with #7126: we could maybe use [send_if_modified](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html#method.send_if_modified) instead? It can't fail and also gives a natural place to check whether the new info is actually a change (and therefore skip the timer rebuild).

File: oximeter/collector/src/agent.rs [hunk 4]
```diff
@@ -131,10 +144,128 @@ async fn perform_collection(
...
+        // producer's information. In the latter case, recreate the future for
+        // the collection itself with the new producer information.
+        let collection_result = 'collection: loop {
+            tokio::select! {
+                biased;
+
+                maybe_update = producer_info_rx.changed() => {
+                    match maybe_update {
+                        Ok(_) => {
+                            let update = producer_info_rx.borrow().clone();
```
Comment (jgallagher) [inline oximeter/collector/src/agent.rs:219]: I think this should be

```suggestion
                            let update = producer_info_rx.borrow_and_update().clone();
```

(and actually, I think nearly _all_ watch channel receivers should use `borrow_and_update` instead of `borrow`; I wish they were flipped and `borrow` did the right thing and `borrow_without_update` was the longer-to-write one).

`borrow_and_update` is needed to correctly handle this ordering:

1. Producer info changes
2. We wake up from `producer_info_rx.changed()` on line 216
3. Producer info changes again
4. We `borrow()` here; we'll see the info from the second change, but borrow doesn't mark the value as seen in the channel. When we `continue 'collection`, `producer_info_rx.changed()` immediately fires again because there's been a change since the last `.changed()` call

File: oximeter/collector/src/agent.rs [hunk 5]
```diff
@@ -656,7 +928,10 @@ impl OximeterAgent {
         //
         // NOTE: This can either mean that the collection completed
         // successfully, or an error occurred in the collection pathway.
-        futures::future::join_all(collection_oneshots).await;
+        futures::future::try_join_all(collection_oneshots)
+            .await
+            .map(|_| ())
+            .map_err(|_| ForcedCollectionQueueFull)
```
Comment (jgallagher) [inline oximeter/collector/src/agent.rs:934]: Do we need a `.map_err` here? I thought the oneshots would get back `ForcedCollectionQueueFull` if that's why they failed.

**Review by jgallagher — APPROVED** [review]
LGTM, thanks! This was quite a bit of work. I'm a little worried we've still missed some edge case, but there's a lot of logging in here so if so, hopefully it's easy to identify.

File: oximeter/collector/src/agent.rs (hunk 2 above)
Comment (bnaecker) [reply oximeter/collector/src/agent.rs:355]: That's true, yeah. I've sort of been operating on the assumption that queues only close if we're shutting down anyway. But we can still send an error of some kind, I'll update this.

File: oximeter/collector/src/agent.rs (hunk 3 above)
Comment (bnaecker) [reply oximeter/collector/src/agent.rs:396]: That sounds good. I'm not familiar with that interface, so I'll need to read up a bit. I was going to merge #7126 first anyway, and so I can make sure to handle conditional modification here too. Thanks for the tip!

File: oximeter/collector/src/agent.rs (hunk 5 above)
Comment (bnaecker) [reply oximeter/collector/src/agent.rs:934]: I switched this from `join_all` to `try_join_all`, which means this is now fallible. I'm not sure that's what we want though, after looking at it again. I want to drive all the futures to completion, and return this error if any one of them failed. `try_join_all` will cancel all the others as soon as it finds an error. I'll rework this a bit to make sure we're doing the right thing here.

**Comment by bnaecker** [comment]
Thanks for the review @jgallagher!

> I'm a little worried we've still missed some edge case

I've been waffling on adding a lot more tests here. The upshot of separating the two tasks out is that it should be easier to test them in isolation. Do you think it's worthwhile to do that? I don't want to bloat the PR, but now that we've seen a few pretty nasty bugs in this file, it seems like a good investment.

File: oximeter/collector/src/agent.rs (hunk 3 above)
Comment (bnaecker) [reply oximeter/collector/src/agent.rs]: Reworked to use `send_if_modified()` in 010e3a273.

File: oximeter/collector/src/agent.rs (hunk 5 above)
Comment (bnaecker) [reply oximeter/collector/src/agent.rs]: I reworked this in 69dfa7a9d. We now return an error enum, rather than an empty struct, to distinguish between failing to send a force-collection request to the collection task because it was full; and failing to send or receive the result because the channel was closed.

File: oximeter/collector/src/agent.rs (hunk 4 above)
Comment (bnaecker) [reply oximeter/collector/src/agent.rs]: Done in 69dfa7a9d, thanks for the careful look.

File: oximeter/collector/src/agent.rs (hunk 2 above)
Comment (bnaecker) [reply oximeter/collector/src/agent.rs]: Added an error variant for this case specifically in 69dfa7a9d.
